nnc/Store.swift

```diff
@@ -147,71 +200,114 @@ private let fpzipDecode:
-        guard status != COMPRESSION_STATUS_ERROR else { return 0 }
-      } while status == COMPRESSION_STATUS_OK && stream.dst_size > 0
-      decodedSize[0] = decodedSize[0] - stream.dst_size
-      return 1
+private let ezm8Decode:
+  @convention(c) (
+    UnsafeRawPointer?, Int, Int32, UnsafePointer<Int32>?, Int32, UInt32, UnsafeMutableRawPointer?,
+    UnsafeMutableRawPointer?, UnsafeMutablePointer<Int>?
+  ) -> Int32 = {
+    data, dataSize, dataType, dimensions, dimensionCount, identifier, context, decoded, decodedSize
+    in
+    guard dataType == Int32(CCV_16F) else { return 0 }
+    guard identifier == 0x511 else { return 0 }
+    guard let data = data, let dimensions = dimensions, let decoded = decoded,
+      let decodedSize = decodedSize, dimensionCount > 0
+    else { return 0 }
+    var floatCount = decodedSize[0] / MemoryLayout<Float16>.size
+    let exponentZipSize = Int(data.assumingMemoryBound(to: Int32.self)[0])
+    let exponentZipData = data.advanced(by: MemoryLayout<Int32>.size)
+    let exponentBuffer = UnsafeMutablePointer<UInt8>.allocate(capacity: floatCount)
+    defer { exponentBuffer.deallocate() }
+    guard unzip(data: exponentZipData,
+                dataSize: exponentZipSize,
+                unzippedData: exponentBuffer,
+                unzippedDataSize: &floatCount) else { return 0 }
+    let floatsWithoutExp = exponentZipData.advanced(by: exponentZipSize).assumingMemoryBound(to: UInt8.self)
+    let decodedAsInts = decoded.assumingMemoryBound(to: UInt16.self)
+    for i in 0..<floatCount {
+      let floatWithoutExp = UInt16(floatsWithoutExp[i])
+      let signBit = floatWithoutExp >> 7
+      let mantissa = (floatWithoutExp & 0x7f) << 3
+      let exponent = UInt16(exponentBuffer[i])
+      decodedAsInts[i] = (signBit << 15) | (exponent << 10) | mantissa
     }
+
+    return 1
+  }
+
+#if canImport(Foundation) && canImport(Compression)
+  import Foundation
+  import Compression
+
+  func zip(data: UnsafeRawPointer, dataSize: Int, zippedData: UnsafeMutableRawPointer, zippedDataSize: UnsafeMutablePointer<Int>) -> Bool {
+    let outputSize = compression_encode_buffer(
+      zippedData.assumingMemoryBound(to: UInt8.self), zippedDataSize[0],
+      data.assumingMemoryBound(to: UInt8.self), dataSize, nil, COMPRESSION_ZLIB)
+    guard outputSize > 0 else { return false }
+    zippedDataSize[0] = outputSize
+    return true
+  }
+
+  private func unzip(data: UnsafeRawPointer, dataSize: Int, unzippedData: UnsafeMutableRawPointer, unzippedDataSize: UnsafeMutablePointer<Int>) -> Bool {
+    let data = NSData(bytesNoCopy: UnsafeMutableRawPointer(mutating: data), length: dataSize, deallocator: .none)
+    guard let decompressedData = try? data.decompressed(using: .zlib) else { return false }
+    guard decompressedData.length <= unzippedDataSize[0] else { return false }
+    unzippedDataSize[0] = decompressedData.length
+    memcpy(unzippedData, decompressedData.bytes, decompressedData.length)
+    return true
+  }
```

Comment:
here i tried using a simpler way of doing decompression, but i can revert if you prefer the previous way

Reply:
Yeah, let's do the previous way. If you run `bazel run test:nnc`, this change shouldn't pass. We have situation that the decompressed data length is larger than the unzippedDataSize[0] (you can read part of the tensor, like the first 2 floats), previous way (using streaming API) can do that, but this cannot.